File: Practica3/Marco teorico/3. OCP/OCP.cpp
#include <iostream>
#include <vector>
using namespace std;

class PaymentMethod
{
public:
   virtual void pay(double amount) = 0;
   virtual ~PaymentMethod() {}
};
// Clase para representar un pago en efectivo
class CashPayment : public PaymentMethod {
public:
 void pay(double amount) override
 {
   cout << "Pagando $" << amount << " en efectivo\n";
 }
};
// Clase para representar un pago con tarjeta de crédito
class CreditCardPayment : public PaymentMethod {
public:
 void pay(double amount) override
 {
   cout << "Pagando $" << amount << " con tarjeta de crédito\n";
 }
};
// Clase para representar una compra en la tienda
class Purchase
{
private:
   double amount;
   PaymentMethod *paymentMethod;
public:
 Purchase(double amount, PaymentMethod *paymentMethod) : amount(amount), paymentMethod(paymentMethod) {}
 void processPayment()
 {
   paymentMethod->pay(amount);
 }
};
// Función para procesar una compra con un tipo de pago determinado
void processPurchase(double amount, PaymentMethod *paymentMethod)
{
   Purchase purchase(amount, paymentMethod);
   purchase.processPayment();
}

int main()
{
   // Crear una compra de $50 con pago en efectivo
   PaymentMethod *cashPayment = new CashPayment();
   processPurchase(50, cashPayment);
   delete cashPayment;
   // Crear una compra de $100 con pago con tarjeta de crédito
   PaymentMethod *creditCardPayment = new CreditCardPayment();
   processPurchase(100, creditCardPayment);
   delete creditCardPayment;
   return 0;
}

File: Practica3/Marco teorico/3. OCP/OCP.py
class PaymentMethod:
    def pay(self, amount):
        pass

class CashPayment(PaymentMethod):
    def pay(self, amount):
        print("Pagando $" + str(amount) + " en efectivo")

class CreditCardPayment(PaymentMethod):
    def pay(self, amount):
        print("Pagando $" + str(amount) + " con tarjeta de crédito")

class Purchase:
    def __init__(self, amount, payment_method):
        self.amount = amount
        self.payment_method = payment_method

    def process_payment(self):
        self.payment_method.pay(self.amount)

def process_purchase(amount, payment_method):
    purchase = Purchase(amount, payment_method)
    purchase.process_payment()


cash_payment = CashPayment()
process_purchase(50, cash_payment)
    
credit_card_payment = CreditCardPayment()
process_purchase(100, credit_card_payment)